### PR TITLE
Show the command line arguments on the cache proxy card

### PR DESCRIPTION
**Pull request by iain-macdonald**
Screenshot 2026-07-22 at 3 34 22 PM

### CODE REVIEW

File: enterprise/server/cache_proxy_registration/cache_proxy_registration.go [hunk 1]
```diff
@@ -148,6 +149,25 @@ func Register(env *real_environment.RealEnv) error {
...
+// their default values are omitted, and the values of secret flags are
+// redacted.
+func configuredFlags() []string {
+	var flags []string
+	flag.CommandLine.VisitAll(func(flg *flag.Flag) {
+		value := flg.Value.String()
+		if value == flg.DefValue {
+			return
+		}
+		if flag.IsSecret(flg) {
```
[the remark this answers is not in the document]
Comment (iain-macdonald) [reply enterprise/server/cache_proxy_registration/cache_proxy_registration.go]: Fixed

File: enterprise/server/cache_proxy_registration/cache_proxy_registration.go [hunk 2]
```diff
@@ -121,6 +121,7 @@ func Register(env *real_environment.RealEnv) error {
 		Labels:               trimmedLabels,
 		AllocatedCpuMillis:   resources.GetAllocatedCPUMillis(),
 		AllocatedMemoryBytes: resources.GetAllocatedRAMBytes(),
+		ConfiguredFlags:      configuredFlags(),
```
Comment (dan-stowell) [inline enterprise/server/cache_proxy_registration/cache_proxy_registration.go]: What happens if the cache proxy reloads config on SIGHUP?

Comment (iain-macdonald) [reply enterprise/server/cache_proxy_registration/cache_proxy_registration.go]: Good catch. As this change was written, it would keep reporting the old flags. I updated it to re-scan (and re-redact) the flags on each heartbeat. That shouldn't be too much overhead and keeps them correct across SIGHUPs.

File: server/util/flag/flag_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+func TestIsSecret(t *testing.T) {
+	for name, want := range map[string]bool{
+		// A plain flag has no secrets.
+		"test.is_secret.plain": false,
+		// A flag declared with the Secret tag.
+		"test.is_secret.secret": true,
+		// A flag whose type has a config:"secret" field.
+		"test.is_secret.credentials": true,
+		// A flag whose type reaches a config:"secret" field via nesting.
+		"test.is_secret.registry": true,
```
Comment (tempoz) [inline server/util/flag/flag_test.go]: Should probably also test the negative cases for structs / slices / maps; as-is, it would be easy to regress by accidentally always redacting any of these.

File: server/util/flag/flag.go [hunk 4]
```diff
@@ -31,6 +34,48 @@ var Args = flag.Args
...
+	}
+	t, err := common.GetTypeForFlagValue(flg.Value)
+	if err != nil {
+		// Unrecognized flag value type; assume it may contain secrets.
+		return true
+	}
+	return typeContainsSecrets(t, map[reflect.Type]bool{})
+}
+
+func typeContainsSecrets(t reflect.Type, seen map[reflect.Type]bool) bool {
```
Comment (tempoz) [inline server/util/flag/flag.go]: This function can be moved to `server/util/redact`.

File: server/util/flag/flag.go [hunk 5]
```diff
@@ -31,6 +34,48 @@ var Args = flag.Args
...
+// IsSecret returns whether the given flag's value may contain secrets,
+// meaning it should never be displayed or logged: either the flag was
+// declared with the Secret tag, or its type contains a struct field tagged
+// config:"secret" (at any depth). Note that for the latter, YAML config
+// export redacts just the secret fields (see flagyaml.RedactSecrets);
+// callers of this coarser check should redact the whole value.
+func IsSecret(flg *flag.Flag) bool {
+	if s, ok := flg.Value.(common.Secretable); ok && s.IsSecret() {
+		return true
+	}
```
Comment (tempoz) [inline server/util/flag/flag.go]: This portion of the function can go in `server/util/flagutil/common` so it can be re-used in `server/util/flagutil/yaml`, and then it can be exposed in `server/util/flagutil` and provide much better ergonomics when checking if a flag has been tagged as `Secret`.

`server/util/flag` is really only for flag type definitions.

File: server/util/flag/flag.go [hunk 6]
```diff
@@ -31,6 +34,48 @@ var Args = flag.Args
 type Flag = flag.Flag
 type FlagSet = flag.FlagSet
 
+// IsSecret returns whether the given flag's value may contain secrets,
+// meaning it should never be displayed or logged: either the flag was
+// declared with the Secret tag, or its type contains a struct field tagged
+// config:"secret" (at any depth). Note that for the latter, YAML config
+// export redacts just the secret fields (see flagyaml.RedactSecrets);
+// callers of this coarser check should redact the whole value.
+func IsSecret(flg *flag.Flag) bool {
```
Comment (tempoz) [inline server/util/flag/flag.go]: This function can be moved to `server/util/redact`. Maggie recently exposed `common.GetTypeForFlagValue` via `flagutil`, so there is no need for this function to be in a flag-specific package, and `server/util/redact` is very flag-focused anyway.

**Review by tempoz — CHANGES_REQUESTED** [review]
Logic looks great; just requested some changes regarding where some of the functions go and tests to cover more negative cases.

File: enterprise/server/cache_proxy_registration/cache_proxy_registration.go (hunk 2 above)
Comment (dan-stowell) [reply enterprise/server/cache_proxy_registration/cache_proxy_registration.go]: Sweet!

File: server/util/flag/flag_test.go (hunk 3 above)
Comment (iain-macdonald) [reply server/util/flag/flag_test.go]: Good idea, added a few negative test cases.

File: server/util/flag/flag.go (hunk 4 above)
Comment (iain-macdonald) [reply server/util/flag/flag.go]: Done

File: server/util/flag/flag.go (hunk 6 above)
Comment (iain-macdonald) [reply server/util/flag/flag.go]: Done

File: server/util/flag/flag.go (hunk 5 above)
Comment (iain-macdonald) [reply server/util/flag/flag.go]: Moved this block to `flagutil.go` (is that what you meant?). Note that I can't use it from `yaml.go` because it introduces this dependency cycle:


```
ERROR: /home/iain/code/buildbuddy/server/util/flag/BUILD:3:11: in go_library rule //server/util/flag:flag: cycle in dependency graph:
    //server/util/authutil:authutil (3ea679ce959b013b08f052a8f278cefba204fdc39fb959bcfddbae75137d04b3)
    //server/util/authutil:authutil (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
.-> //server/util/flag:flag (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
|   //server/util/flagutil/types/autoflags/tags:tags (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
|   //server/util/flagutil/yaml:yaml (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
|   //server/util/flagutil:flagutil (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
`-- //server/util/flag:flag (778cd499bd8a5acf7efc7bef0e7812ca70115265611a8a1791451db2908900c8)
```

Comment (tempoz) [reply server/util/flag/flag.go]: no, I meant moving it to `server/util/flagutil/common/common.go` and then exposing it as an API to the non-flagutil code by wrapping it in `server/util/flagutil/flagutil.go`

**Review by tempoz — CHANGES_REQUESTED** [review]

File: server/util/flag/flag.go (hunk 5 above)
Comment (iain-macdonald) [reply server/util/flag/flag.go]: Ahhh, got it. Done.

**Review by tempoz — APPROVED** [review]
